front/src/applications/operationalStudies/components/MacroEditor/utils.ts

```diff
@@ -238,3 +242,38 @@ export const getTrainrunFrequencyFromTimetableItem = (
   }
   return trainrunFrequency;
 };
+
+export const getNetzgrafikColors = () =>
+  Object.entries(NETZGRAFIK_COLOR_PALETTE).map(([colorRef, baseColor], index) => {
+    const colorVariant = CATEGORY_COLOR_VARIANTS[baseColor];
+
+    return {
+      id: index,
+      colorRef,
+      color: colorVariant[50],
+      colorFocus: colorVariant[70],
+      colorMuted: colorVariant[10],
+      colorRelated: colorVariant[50],
+      colorDarkMode: colorVariant[50],
+      colorDarkModeFocus: colorVariant[70],
+      colorDarkModeMuted: colorVariant[10],
+      colorDarkModeRelated: colorVariant[50],
+    };
+  });
+
+export const getTrainCategoryFromId = (id: number): TrainCategory | null => {
+  if (id === OSRD_TRAINRUN_CATEGORY_MAPPING.NO_CATEGORY.id) {
+    return null;
+  }
+
+  for (const key of Object.keys(OSRD_TRAINRUN_CATEGORY_MAPPING) as (
+    | TrainCategory
+    | 'NO_CATEGORY'
+  )[]) {
```

Comment:
Here you have to use a cast because Record is not strongly typed: ts thinks that `Object.keys(OSRD_TRAINRUN_CATEGORY_MAPPING)` are `string[]` even if you have precised that they are `TrainCategory | 'NO_CATEGORY'`.

To help typescript, it's better to use Maps than Records. You can create a Map with `const a = new Map()`, then set its values `a.set(key, value)` and finally retrieve its keys correctly typed with `a.keys()`
```suggestion
  for (const key of OSRD_TRAINRUN_CATEGORY_MAPPING.keys()) {
```

PS: you can get a map value by key with `const value = a.get(key)`. For your information, `value` might always be undefined but you can add some `!` where needed

Reply:
Thanks for the proposition !
Done